language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/Primitive.scala

```diff
@@ -47,9 +47,12 @@ sealed abstract class Primitive {
   type Optional[+A] = scala.Option[A]
   val Optional: scala.Option.type = scala.Option
 
-  type Map[+A] = imm.Map[String, A]
+  type Map[+V] = imm.Map[String, V]
   val Map: imm.Map.type = imm.Map
 
+  type GenMap[K, +V] = imm.Map[K, V]
+  val GenMap: imm.Map.type = imm.Map
```

Comment:
It's fine to permit builtin map key treatment under a rather big *if*: that for all ⊢ₛ τ, e₁: τ, e₂: τ, `e₁ ~ᵥ e₂` ≡ `e₁ₛ == e₂ₛ`, where eₙₛ are the Scala codegen representations of eₙ respectively, and `==` is the builtin Scala equality method.

My reading of ~ᵥ tells me this is currently true. However, if it is not, or it becomes untrue (such as by introduction of key comparison based on DAML evaluation), then `imm.Map` cannot be used here for that reason. At the minimum in such a case, you would have to define a typeclass for ~ᵥ (and derive it for generated records and variants as well), and that is assuming that would even work.

----

However, there's another reason that `imm.Map` may be wrong here. One can currently assume that if you decode a value into Scala codegen, then encode it and submit it as a create or choice argument, it will be equal to the original decoded value. But the spec says "The keys are returned by first-insertion order"; that would be scrambled by most choices that fit `imm.Map` here.

Reply:
¹ Here I mean in the strongest, substitution sense: ∀a,b,f: a = b ⇒ f(a) = f(b)

Reply:
Thanks @S11001001. You have made a very good point. 
I replace the `imm.Map` by `InsertOrdMap`.

Reply:
By the way, your "big _if_" is an **important** assumption of the equality we define for map key. 
- No user defined equality.
- Equality is reflexive for all serializable values.

The equality of `Map`  is still independent of the entries ordering.